### PR TITLE
add dynamic reload policy lines when authorization policy file is modified

### PR DESCRIPTION
**Pull request by chaolou**
hi:
the existing authorization can't auto reload the policy file.  When somebody want to modify the policy strategy, he must restart apiServer after modifying the authorization config file.

This PR add a backend check goroutine when apiServer initiallize, and judge config whether has being modified by comparison file's modify time every 5 seconds,

### CODE REVIEW

**Comment by chaolou** [comment]
I signed it!

File: pkg/apiserver/authz.go [hunk 1]
```diff
@@ -108,3 +115,36 @@ func NewAuthorizerFromAuthorizationConfig(authorizationModes []string, authoriza
...
+					return
+				}
+
+				glog.Infof("authorizationPolicyFile: %s is modified, reload finished, number of policy lines change from %d to %d", file, len(*policyAuthz.PL), len(newPolicyList))
+				policyAuthz.Lock.Lock()
+				*(policyAuthz.PL) = newPolicyList
+				policyAuthz.Lock.Unlock()
+			}
+		}()
+		time.Sleep(5 * time.Second)
```
Comment (mqliang) [inline pkg/apiserver/authz.go]: It' better to define sleep time as a constant. Such as `const recheckIntervel = 5*time.Second`

**Comment by smarterclayton** [comment]
So in general we don't auto reload these things inside the server today, because layers above the authorizer may cache them. Any change here needs to be well scoped to prevent that abstraction from leaking.

@deads2k

File: pkg/apiserver/authz.go [hunk 2]
```diff
@@ -92,10 +97,12 @@ func NewAuthorizerFromAuthorizationConfig(authorizationModes []string, authoriza
 				return nil, errors.New("ABAC's authorization policy file not passed")
 			}
 			abacAuthorizer, err := abac.NewFromFile(authorizationPolicyFile)
+			authorizerLock := &abac.PolicyAuthorizer{PL: &abacAuthorizer, Lock: &sync.RWMutex{}}
```
Comment (smarterclayton) [inline pkg/apiserver/authz.go]: Whether this reloads or not needs to be configurable, not always true.

For instance, when running apiserver in HA configurations, administrators may not want this behavior.

**Comment by smarterclayton** [comment]
I think this needs to be a different authorizer implementation (or wrapper) that is clearly decoupled from the file based authorizer code.

**Comment by chaolou** [comment]
Thank you for response promptly.
But,  I want to consult you some questions
1. I don't understand what meaning about you said `because layers above the authorizer may cache them. Any change here needs to be well scoped to prevent that abstraction from leaking.`
I don't find any above cache for authorizer in apiserver, this reload policy method can ensure the `Authorize` use the new load policies by passing by pointer.

2. Now `the file based authorizer` is the only config style supported for authorization, maybe another style by adding authorization policy into etcd can solve the exsit problem which can't change authorization policy dynamic. I want to known whether your have already some decided enhance plans for this.

Of course, adding a control flag denote whether need to reload policies or not should be necessary.

**Comment by deads2k** [comment]
> I don't understand what meaning about you said because layers above the authorizer may cache them. Any change here needs to be well scoped to prevent that abstraction from leaking.
> I don't find any above cache for authorizer in apiserver, this reload policy method can ensure the Authorize use the new load policies by passing by pointer.

A caching authorizer is a common next step that we found we needed very quickly in OpenShift.  We don't want to make that sort of a growth here more difficult.

**Comment by deads2k** [comment]
> Now the file based authorizer is the only config style supported for authorization, maybe another style by config authorization policy into etcd as a resource then apiserver watch it(just like kubelet's SourceApiserver) can solve the exsisting problem which can't change authorization policy dynamic. 
> I really want to known whether your have already some decided enhance plans for this.

Take a look at https://github.com/kubernetes/kubernetes/pull/12209 for a starting point and https://github.com/openshift/origin/blob/master/docs/proposals/policy.md for a coarse and slightly out of date view into what we've done downstream.

**Comment by deads2k** [comment]
@kubernetes/kube-iam

**Pull request closed without merging**